### PR TITLE
feat: Version Compatibility Based on Gem Specs

### PR DESCRIPTION
**Pull request by arielvalentin**
This changes provides instrumentation authors the ability to use gem specifications for version compatibility checks instead of having to write a custom compatibility block.

In order to take advantage of this feature instrumentation authors must:

1. Declare the library and version constraints as a development dependency
2. Use instrumentation naming conventions e.g. `opentelemetry-instrumentation-ruby_kafka`

Related https://github.com/open-telemetry/opentelemetry-ruby/issues/988

Some Concerns:

- There are certainly many more test scenarios to take into consideration. E.g. When multiple versions of the gem are installed and not yet activated.  
- Naming is strict and requires that the instrumentation name suffix matches the target library. We currently have some cases where the instrumentation names do not match e.g. `ruby-kafka` and `activesupport`

### CODE REVIEW

File: instrumentation/base/lib/opentelemetry/instrumentation/base.rb [hunk 1]
```diff
@@ -293,11 +293,15 @@ def enabled?(config = nil)
...
-        dependency = instrumentation_spec&.development_dependencies&.find { |spec| spec.name == library_spec.name }
-        dependency&.requirement&.satisfied_by?(library_spec.version) == true
+        begin
+          instrumentation_spec = Gem.loaded_specs[instrumentation_gem_name] || Gem::Specification.find_by_name(instrumentation_gem_name)
+          library_spec = Gem.loaded_specs[library_name] || Gem::Specification.find_by_name(library_name)
+
+          dependency = instrumentation_spec.development_dependencies.find { |spec| spec.name == library_spec.name }
+          dependency&.requirement&.satisfied_by?(library_spec.version) == true
+        rescue Gem::MissingSpecError
+          return false
```
Comment (ericmustin) [inline instrumentation/base/lib/opentelemetry/instrumentation/base.rb:303]: A debug or warn log here couldn’t hurt?

Comment (arielvalentin) [reply instrumentation/base/lib/opentelemetry/instrumentation/base.rb:303]: Sure thing.

Comment (arielvalentin) [reply instrumentation/base/lib/opentelemetry/instrumentation/base.rb]: I chose to use `handle_error`. Seem reasonable?

**Comment by arielvalentin** [comment]
Ping @open-telemetry/ruby-maintainers @open-telemetry/ruby-approvers 🙏

**Comment by ericmustin** [comment]
@arielvalentin My two cents is i think this looks great and is an improvement on the current system. It feels _ a little_ hacky to me to rely on development dependancies in a gemspec, for something like an instrumentation that doesn't necessarily need to be packaged as a gem to begin with, but since this doesn't override the compatible block, but only acts as a useful fallback/default, I don't see the harm. that being said I'd like some other folks to chime in here, I brought this up in the SIG, would be keen to hear other POVs here.

File: instrumentation/base/lib/opentelemetry/instrumentation/base.rb (hunk 1 above)
Comment (ericmustin) [reply instrumentation/base/lib/opentelemetry/instrumentation/base.rb]: 👍

**Comment by arielvalentin** [comment]
> @arielvalentin My two cents is i think this looks great and is an improvement on the current system. It feels _ a little_ hacky to me to rely on development dependancies in a gemspec, for something like an instrumentation that doesn't necessarily need to be packaged as a gem to begin with, but since this doesn't override the compatible block, but only acts as a useful fallback/default, I don't see the harm. that being said I'd like some other folks to chime in here, I brought this up in the SIG, would be keen to hear other POVs here.

@ericmustin I would think folks that are writing their own auto-instrumentation that are not packaged as a gem are likely either implemented first party or patched in their applications. It's likely they are aware what the compatibility requirements are in their own projects or may not be using instrumentation at all. 

Could you elaborate a bit on what you think makes it _a little hacky_?

**Comment by ericmustin** [comment]
> Could you elaborate a bit on what you think makes it a little hacky?

I don't think that checking [development dependancies in a gemspec from production code](https://github.com/open-telemetry/opentelemetry-ruby/pull/1025/files#diff-7379de806670e7a913ebbe1d8e27945a6367cc83c94d79e55434de15cdd16732R316) is a common convention. Maybe hacky isn't the right term.  That being said I think it's fine imo, just, unusual. Again i'd like some other input here from @open-telemetry/ruby-maintainers @open-telemetry/ruby-approvers

**Comment by arielvalentin** [comment]
@robertlaurin do you have any time this week to take a look?

**Comment by arielvalentin** [comment]
Closing per my offline conversation with @robertlaurin https://github.com/open-telemetry/opentelemetry-ruby/issues/988#issuecomment-1018676853

**Pull request closed without merging**